### ISSUE TITLE
#[from] without implicit #[source]

### ISSUE THREAD

**Issue by CodeSandwich** [dtolnay/thiserror#74]
### TL;DR
It would be great to have an equivalent of `#[from]` attribute, which doesn't provide implicitly the functionality of `#[source]`. 
### The setting
I've got an enum:
```rust
#[derive(Error)]
enum MyError {
    Failure(String),
}
```
and a function I didn't write:
```rust
fn fallible() -> Result<(), String> { ... }
```
which I want to use inside my function:
```rust
fn meta_fallible() -> Result<(), MyError> {
    ...
    fallible()?;
    ...
}
```
### The problem
This won't compile and AFAIK there's no ergonomic way to make `MyError` smart enough to make it work.
Right now `meta_fallible` won't be able to convert error `String` to `MyError`, because it lacks right `from`. Let's try to add it:
```rust
#[derive(Error)]
enum MyError {
    Failure(#[from] String),
}
```
Now it won't compile, because while `From<String> for MyError` would be fine, `String` can't be used as `&dyn std::error::Error`, so it will fail as a thing returned from `source`. The only way to bypass this is manual implementation of `From<String> for MyError`.

**Comment by dtolnay** [comment]
This is intentional, because I would prefer to stick with being an `Error` derive and not morph into a general purpose `From` derive. Consider pulling in a different macro to generate this `From` impl, or stick with handwriting it. Or better: send a PR to the crate containing `fallible` to implement Error for its error types.

**Comment by alilleybrinker** [comment]
One pattern you can use is to make a wrapper error type that actually implements `Error`, like so:

```rust
#[derive(thiserror::Error)]
enum MyError {
    #[error("{0}")]
    Failure(#[from] FallibleError),
}

#[derive(thiserror::Error)]
enum FallibleError {
    #[error("{0}")]
    Msg(String),
}
```

And then at the call site for `fallible`:

```rust
fallible().map_err(FallibleError::Msg)?;
```
